### ISSUE TITLE
Getting started with type hints

### ISSUE THREAD

**Issue by iancze** [MPoL-dev/MPoL#54]
**Is your feature request related to a problem or opportunity? Please describe.**
It might be worthwhile to slowly start adding type hints to some of the core routines. In theory this should improve reliability.

More information [here](https://mypy.readthedocs.io/en/stable/introduction.html) and an introduction [here](https://bernat.tech/posts/the-state-of-type-hints-in-python/).

**Describe the solution you'd like**
It seems like this is something we can start doing [gradually](https://mypy.readthedocs.io/en/stable/existing_code.html). Probably worth starting with files in `src/mpol/constants.py` or `src/mpol/utils.py`.

- [ ] We probably need to mirror some of the PyTorch [mypy.ini](https://github.com/pytorch/pytorch/blob/master/mypy.ini) file.
- [ ] Add type checking to continuous integration on Github Actions

**Comment by kadri-nizam** [comment]
Will track files that have been type hinted here.

- [ ] connectors.py
- [x] constants.py
- [ ] coordinates.py
- [ ] datasets.py
- [ ] fourier.py
- [ ] gridding.py
- [ ] images.py
- [ ] losses.py
- [ ] precomposed.py
- [ ] spheroidal_gridding.py
- [ ] utils.py

**Comment by iancze** [comment]
For `torch.Tensor`, we'd like to get more specific with specific dtypes and shapes. 

E.g., for a loss function like

```
def chi_squared(model_vis, data_vis, weight):
    ....
```

we'd like to specify that all of the input tensors are the same shape (could be 1D, 2D, or more, but must be the same shape), `model_vis` and `data_vis` are complex vectors, `weight` is a float type, and the return value is a float. I'm used to being able to easily type something like this in Julia ([e.g.,](https://github.com/iancze/DiskJockey/blob/main/src/image.jl#L48)) but it seems like this is not really standardized yet with PyTorch Tensors and only recently possible with Numpy.

I found discussions like [this](https://github.com/pytorch/pytorch/issues/73359), [this](https://github.com/pytorch/pytorch/issues/98702), and [this](https://github.com/pytorch/pytorch/issues/98702), but it seems like the capability to do this has been slowly evolving. There are third party tools like [jaxtyping](https://github.com/google/jaxtyping) that provide support for PyTorch typing (despite the name), but I'm hesitant to bring in additional third party dependencies (that also have their own Python version dependencies). 

Right now, for PyTorch the best I think we can do is

```
def chi_squared(
    model_vis: torch.Tensor, data_vis: torch.Tensor, weight: torch.Tensor
) -> torch.Tensor:
```

which at least enforces that input and output should be of type `torch.Tensor`, but does nothing to enforce dtypes or shapes. And then the plan is to adopt more sophisticated type checking functionality for Tensor dtypes and shapes as it makes its way into standard packages (which seems likely to happen before too long).
